sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceV2Strategy.scala

```diff
@@ -46,25 +46,30 @@ object DataSourceV2Strategy extends Strategy with PredicateHelper {
       case r: SupportsPushDownFilters =>
         // A map from translated data source filters to original catalyst filter expressions.
         val translatedFilterToExpr = mutable.HashMap.empty[sources.Filter, Expression]
+        val translatedFilters = mutable.ArrayBuffer.empty[sources.Filter]
         // Catalyst filter expression that can't be translated to data source filters.
         val untranslatableExprs = mutable.ArrayBuffer.empty[Expression]
 
         for (filterExpr <- filters) {
-          val translated = DataSourceStrategy.translateFilter(filterExpr)
-          if (translated.isDefined) {
-            translatedFilterToExpr(translated.get) = filterExpr
-          } else {
+          val translated =
```

Comment:
With partial filter push down in Or operator, the result of pushedFilters() might not exist in the mapping translatedFilterToExpr. To fix it, this PR changes the mapping translatedFilterToExpr as leaf filter expression to sources.filter, and later on rebuild the whole expression with the mapping.

Reply:
Yep. Actually, when I tested your PR before, I also notice that. The new `Or` doesn't work in the end-to-end case. That's the reason to ask this. I'll test more.

Thank you for making this PR working, @gengliangwang !

Reply:
BTW, it would be great to add a real `comment` in the code. It's non-trivial to figure out.

Reply:
Thanks for for the testing. Appreciate it!